File: weyl-core/src/main/java/io/quartic/weyl/core/compute/SpatialJoiner.java
package io.quartic.weyl.core.compute;

import io.quartic.common.SweetStyle;
import io.quartic.weyl.core.model.Feature;
import io.quartic.weyl.core.model.IndexedFeature;
import io.quartic.weyl.core.model.Layer;
import org.immutables.value.Value;

import java.util.stream.Stream;

public class SpatialJoiner {
    @SweetStyle
    @Value.Immutable
    public interface Tuple {
        Feature left();
        Feature right();
    }

    @SuppressWarnings("unchecked")
    public Stream<Tuple> innerJoin(Layer leftLayer, Layer rightLayer, SpatialPredicate predicate) {
       return leftLayer.indexedFeatures().parallelStream()
                .flatMap(left -> rightLayer.spatialIndex()
                        .query(left.feature().geometry().getEnvelopeInternal())
                        .stream()
                        .filter(o -> predicate.test(left.preparedGeometry(), ((IndexedFeature) o).feature().geometry()))
                        .map(o -> TupleImpl.of(left.feature(), ((IndexedFeature) o).feature())));
    }
}

File: weyl-core/src/main/java/io/quartic/weyl/core/compute/SpatialJoiner.kt
package io.quartic.weyl.core.compute

import io.quartic.weyl.core.model.Feature
import io.quartic.weyl.core.model.IndexedFeature
import io.quartic.weyl.core.model.Layer
import java.util.stream.Stream

class SpatialJoiner {
    // TODO: tuple concept probably already exists natively in Kotlin
    data class Tuple(val left: Feature, val right: Feature)

    fun innerJoin(leftLayer: Layer, rightLayer: Layer, predicate: SpatialPredicate): Stream<Tuple> {
        return leftLayer.indexedFeatures.parallelStream()
                .flatMap { (preparedGeometry, feature) ->
                    rightLayer.spatialIndex
                            .query(feature.geometry.envelopeInternal)
                            .stream()
                            .filter { o -> predicate.test(preparedGeometry, (o as IndexedFeature).feature.geometry) }
                            .map { o -> Tuple(feature, (o as IndexedFeature).feature) }
                }
    }
}
